File: kirimel/http_client.py
"""
HTTP Client for KiriMel API
"""
import os
import time
import logging
from typing import Optional, Dict, Any, List
import requests

from .exceptions import ApiException, AuthenticationException, RateLimitException, ValidationException


class HttpClient:
    """HTTP client for making API requests"""

    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: str = "https://kirimel.com/api",
        timeout: int = 30,
        retries: int = 3,
    ):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key or os.getenv("KIRIMEL_API_KEY")
        self.timeout = timeout
        self.retries = retries
        self.session = requests.Session()
        self.logger = logging.getLogger(__name__)

    def set_logger(self, logger: logging.Logger) -> None:
        """Set a custom logger"""
        self.logger = logger

    def get(self, path: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Make a GET request"""
        url = self._build_url(path, params or {})
        return self._request("GET", url)

    def post(self, path: str, data: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Make a POST request"""
        return self._request("POST", self._build_url(path), data)

    def put(self, path: str, data: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Make a PUT request"""
        return self._request("PUT", self._build_url(path), data)

    def delete(self, path: str) -> Dict[str, Any]:
        """Make a DELETE request"""
        return self._request("DELETE", self._build_url(path))

    def _build_url(self, path: str, params: Optional[Dict[str, Any]] = None) -> str:
        """Build URL with query parameters"""
        url = f"{self.base_url}/{path.lstrip('/')}"
        if params:
            import urllib.parse
            query_string = urllib.parse.urlencode(params)
            url = f"{url}?{query_string}"
        return url

    def _request(
        self,
        method: str,
        url: str,
        data: Optional[Dict[str, Any]] = None,
        attempt: int = 0,
    ) -> Dict[str, Any]:
        """Make HTTP request with retry logic"""
        self.logger.debug(f"Making {method} request to {url}", extra={"attempt": attempt + 1})

        headers = self._build_headers()

        try:
            response = self.session.request(
                method=method,
                url=url,
                json=data,
                headers=headers,
                timeout=self.timeout,
            )
        except requests.RequestException as e:
            # Network error - retry if attempts remain
            if attempt < self.retries - 1:
                self.logger.warning(f"Network error, retrying...", extra={"error": str(e)})
                time.sleep(0.1 * (attempt + 1))  # Exponential backoff
                return self._request(method, url, data, attempt + 1)
            raise ApiException(f"Network error: {str(e)}")

        if response.status_code >= 400:
            self._handle_error(response, url, attempt)

        return response.json()

    def _build_headers(self) -> Dict[str, str]:
        """Build request headers"""
        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            "User-Agent": "KiriMel-Python-SDK/0.1.0",
        }

        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"

        return headers

    def _handle_error(self, response: requests.Response, url: str, attempt: int) -> None:
        """Handle API errors"""
        try:
            data = response.json()
            message = data.get("message", "API request failed")
            errors = data.get("errors")
        except ValueError:
            message = response.text or "API request failed"
            errors = None

        if response.status_code == 401:
            raise AuthenticationException(message, response.status_code, errors)
        elif response.status_code == 429:
            retry_after = data.get("retry_after") if errors else None
            if retry_after and attempt < self.retries - 1:
                self.logger.info(f"Rate limited, waiting {retry_after}s...")
                time.sleep(retry_after)
                return  # Will retry
            raise RateLimitException(message, response.status_code, errors, retry_after)
        elif response.status_code == 422:
            raise ValidationException(message, response.status_code, errors)
        else:
            raise ApiException(message, response.status_code, errors)

File: src/HttpClient.php
<?php

declare(strict_types=1);

namespace KiriMel;

use KiriMel\Exceptions\ApiException;
use KiriMel\Exceptions\AuthenticationException;
use KiriMel\Exceptions\RateLimitException;
use KiriMel\Exceptions\ValidationException;

/**
 * HTTP Client for making API requests
 */
class HttpClient
{
    private string $baseUrl;
    private ?string $apiKey;
    private int $timeout;
    private int $retries;
    private ?object $logger = null;

    public function __construct(array $config = [])
    {
        $this->baseUrl = rtrim($config['base_url'] ?? 'https://kirimel.com', '/');
        $this->apiKey = $config['api_key'] ?? $this->getEnvApiKey();
        $this->timeout = $config['timeout'] ?? 30;
        $this->retries = $config['retries'] ?? 3;
        $this->logger = $config['logger'] ?? null;
    }

    private function getEnvApiKey(): ?string
    {
        return $_ENV['KIRIMEL_API_KEY'] ?? null;
    }

    public function setLogger(object $logger): void
    {
        $this->logger = $logger;
    }

    /**
     * Make a GET request
     */
    public function get(string $path, array $params = []): array
    {
        $url = $this->buildUrl($path, $params);
        return $this->request('GET', $url);
    }

    /**
     * Make a POST request
     */
    public function post(string $path, array $data = []): array
    {
        return $this->request('POST', $this->buildUrl($path), $data);
    }

    /**
     * Make a PUT request
     */
    public function put(string $path, array $data = []): array
    {
        return $this->request('PUT', $this->buildUrl($path), $data);
    }

    /**
     * Make a DELETE request
     */
    public function delete(string $path): array
    {
        return $this->request('DELETE', $this->buildUrl($path));
    }

    /**
     * Build URL with query parameters
     */
    private function buildUrl(string $path, array $params = []): string
    {
        $url = $this->baseUrl . '/' . ltrim($path, '/');

        if (!empty($params)) {
            $url .= '?' . http_build_query($params);
        }

        return $url;
    }

    /**
     * Make HTTP request with retry logic
     */
    private function request(string $method, string $url, ?array $data = null, int $attempt = 0): array
    {
        $this->log('debug', "Making {$method} request to {$url}", ['attempt' => $attempt + 1]);

        $ch = curl_init();

        curl_setopt_array($ch, [
            CURLOPT_URL => $url,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => $this->timeout,
            CURLOPT_CUSTOMREQUEST => $method,
            CURLOPT_HTTPHEADER => $this->buildHeaders(),
        ]);

        if ($data !== null) {
            curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));
        }

        $response = curl_exec($ch);
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $curlError = curl_error($ch);
        curl_close($ch);

        if ($curlError) {
            // Network error - retry if attempts remain
            if ($attempt < $this->retries - 1) {
                $this->log('warning', "Network error, retrying...", ['error' => $curlError]);
                usleep(100000 * ($attempt + 1)); // Exponential backoff
                return $this->request($method, $url, $data, $attempt + 1);
            }
            throw new ApiException("Network error: {$curlError}", 0);
        }

        $decoded = json_decode($response, true);

        if ($httpCode >= 400) {
            $this->handleError($httpCode, $decoded, $url, $attempt);
        }

        return $decoded;
    }

    /**
     * Build request headers
     */
    private function buildHeaders(): array
    {
        $headers = [
            'Content-Type: application/json',
            'Accept: application/json',
            'User-Agent: KiriMel-PHP-SDK/1.0.0',
        ];

        if ($this->apiKey) {
            $headers[] = "Authorization: Bearer {$this->apiKey}";
        }

        return $headers;
    }

    /**
     * Handle API errors
     */
    private function handleError(int $httpCode, ?array $response, string $url, int $attempt): void
    {
        $message = $response['message'] ?? 'API request failed';
        $errors = $response['errors'] ?? null;

        switch ($httpCode) {
            case 401:
                throw new AuthenticationException($message, 0, $httpCode, $errors);
            case 429:
                $retryAfter = $response['retry_after'] ?? null;
                if ($retryAfter && $attempt < $this->retries - 1) {
                    $this->log('info', "Rate limited, waiting {$retryAfter}s...");
                    sleep($retryAfter);
                    return; // Will retry
                }
                throw new RateLimitException($message, 0, $httpCode, $errors, $retryAfter);
            case 422:
                throw new ValidationException($message, 0, $httpCode, $errors);
            default:
                throw new ApiException($message, 0, $httpCode, $errors);
        }
    }

    /**
     * Log message
     */
    private function log(string $level, string $message, array $context = []): void
    {
        if ($this->logger && method_exists($this->logger, $level)) {
            $this->logger->$level($message, $context);
        }
    }
}
